### ISSUE TITLE
'The following modules were required before honeycomb-beeline' even though beeline was require-ed first

### ISSUE THREAD

**Issue by nathanleclaire** [honeycombio/beeline-nodejs#58]
If I try to add the NodeJS beeline to https://github.com/developit/express-es6-rest-api - I get an error message:

```
The following modules were required before honeycomb-beeline: express
These modules will not be instrumented.  Please ensure honeycomb-beeline is required first.
```

But the `src/index.js` looks like this:

```javascript
require('honeycomb-beeline')({
    writeKey: '<REDACTED>',
    datasetName: 'nathanleclaire.examples-node-beeline',
});
import http from 'http';
import express from 'express';
import cors from 'cors';
import morgan from 'morgan';
import bodyParser from 'body-parser';
import initializeDb from './db';
import middleware from './middleware';
import api from './api';
import config from './config.json';
```

This should work, right?

**Comment by toshok** [comment]
unfortunately not. it depends on the way babel transforms the code.  babel converts the `import`s to `require`s and then hoists them all to the top of the file (above all the executable code, including requires that exist in the source):

```
'use strict';
  
Object.defineProperty(exports, "__esModule", {
        value: true
});

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _express = require('express');

var _express2 = _interopRequireDefault(_express);

.....

require("honeycomb-beeline")({
        datasetName: 'nathanleclaire.examples-node-beeline'
});

```

that error message is ... not exactly right, either.  the require can happen at any point, but that _call_ to configure the beeline has to happen before express is required.

I think the only way to really fix this is to put the beeline require in a separate file, and `require` the index.js file from there.

**Comment by toshok** [comment]
There are a couple of ways to fix this.  The easiest in terms of code is to introduce another .js file (call it `src/honeycomb.js` just for kicks) containing the following:

```javascript
require('honeycomb-beeline')({
    writeKey: '<REDACTED>',
    datasetName: 'nathanleclaire.examples-node-beeline',
});

require('.');
```

Then add a couple of scripts to package.json:

```diff
diff --git a/package.json b/package.json
index 2bde82d..1e63a41 100644
--- a/package.json
+++ b/package.json
@@ -5,8 +5,10 @@
   "main": "dist",
   "scripts": {
     "dev": "nodemon -w src --exec \"babel-node src --presets es2015,stage-0\"",
+    "dev-honeycomb": "nodemon -w src --exec \"babel-node src/honeycomb.js --presets es2015,stage-0\"",
     "build": "babel src -s -D -d dist --presets es2015,stage-0",
     "start": "node dist",
+    "start-honeycomb": "node dist/honeycomb.js",
     "prestart": "npm run -s build",
     "test": "eslint src"
   },
```

then either run:

```
$ npm run dev-honeycomb
```

or:

```
$ npm run build
$ npm run start-honeycomb
```

should work.

Disappointing about that import->require reordering

**Comment by nathanleclaire** [comment]
Bummer that it doesn't work out of the box, but glad there's a workaround! Thanks for the prompt feedback @toshok.

**Comment by jermeo** [comment]
Hello, I have the same problem with a pure js code with  beeline 1.8.1

_The following modules were required before honeycomb-beeline: bluebird,express
These modules will not be instrumented.  Please ensure honeycomb-beeline is required first._

I also made another js file that is called first, but same error.

My code is launch with the node command.

Thanks

**Comment by nhbond** [comment]
> Hello, I have the same problem with a pure js code with beeline 1.8.1
> 
> _The following modules were required before honeycomb-beeline: bluebird,express These modules will not be instrumented. Please ensure honeycomb-beeline is required first._
> 
> I also made another js file that is called first, but same error.
> 
> My code is launch with the node command.
> 
> Thanks
I'm having the same problem.
Were you able to solve this issue?

**Comment by nhbond** [comment]
> Hello, I have the same problem with a pure js code with beeline 1.8.1
> 
> _The following modules were required before honeycomb-beeline: bluebird,express These modules will not be instrumented. Please ensure honeycomb-beeline is required first._
> 
> I also made another js file that is called first, but same error.
> 
> My code is launch with the node command.
> 
> Thanks

Hey @jermeo 
I'm having the same problem.
Were you able to solve this issue?

**Comment by sw-yx** [comment]
chiming in - i'm doing this in a serverless function and its giving me `The following modules were required before honeycomb-beeline: express` error - no es6 modules involved

```js
exports.handler = (event, context, cb) => {
  require("honeycomb-beeline")({
    writeKey: process.env.HONEYCOMB_API_KEY,
    dataset: process.env.HONEYCOMB_DATASET_NAME,
    // ... additional optional configuration ...
  });

  const express = require("express");
  const serverless = require("serverless-http");
  const bodyParser = require("body-parser");
  const router = express.Router();
  const app = express();
  router.get("/", (req, res) => {
    res.json({ postBody: req.body });
  });
  app.use(bodyParser.json());
  app.use("/.netlify/functions/honeycomb", router);
  return serverless(app)(event, context, cb);
};
```

no amount of rearranging works. this also does not work (this is what i tried originally):


```js
  require("honeycomb-beeline")({
    writeKey: process.env.HONEYCOMB_API_KEY,
    dataset: process.env.HONEYCOMB_DATASET_NAME,
    // ... additional optional configuration ...
  });

  const express = require("express");
  const serverless = require("serverless-http");
  const bodyParser = require("body-parser");
  const router = express.Router();
  const app = express();
exports.handler = (event, context, cb) => {
  router.get("/", (req, res) => {
    res.json({ postBody: req.body });
  });
  app.use(bodyParser.json());
  app.use("/.netlify/functions/honeycomb", router);
  return serverless(app)(event, context, cb);
};
```

**Comment by mehdi-khan** [comment]
@sw-yx I'm facing the same issue with a cloud function. Did you find a workaround for this?

**Comment by sw-yx** [comment]
no i did not sorry. i ended up dropping beeline altogether. for the record i was using netlify functions, which uses aws lambda under the hood

**Comment by MikeGoldsmith** [comment]
We are trimming our OSS backlog. We will be closing this issue as it is a low priority for us to fix. It is unlikely that we'll ever get to it, and so we'd like to set expectations accordingly.

If this issue is important to you, please feel free to ping here and we can discuss/re-open.
